proposals/route-blinding.md

```diff
@@ -0,0 +1,242 @@
+payload for `N(0)`.
+
+When `N(0)` receives the onion and decrypts it, it finds `E(0)` in the payload and is able to
+compute the following:
+
+```text
+  ss(0) = H(k(0) * E(0))
+  rho(0) = HMAC256("rho", ss(0))
+  E(1) = H(E(0) || ss(0)) * E(0)
+```
+
+It uses `rho(0)` to decrypt the `encrypted_blob(0)` and discover the `scid` to forward to.
+It forwards the onion to the next node and includes `E(1)` in a TLV field in the message
+extension.
+
+All the following intermediate nodes `N(i)` do the following steps:
+
+```text
+  E(i) <- extracted from TLV extension
+  ss(i) = H(k(i) * E(i))
+  b(i) = HMAC256("blinded_node_id", ss(i)) * k(i)
+  Use b(i) to decrypt the incoming onion
+  rho(i) = HMAC256("rho", ss(i))
+  Use rho(i) to decrypt the `encrypted_blob` inside the onion and discover the next node
+  E(i+1) = H(E(i) || ss(i)) * E(i)
+  Forward the onion to the next node and include E(i+1) in a TLV field in the message extension
+```
+
+### Receiving from a blinded route
+
+When `N(r)` receives the onion message and `E(r)` in the TLV extension, she does the same
+unwrapping as intermediate nodes. The difference is that the onion will be a final onion.
+
+### Unblinding channels via fee probing
+
+The fees and cltv for the blinded route can be abused by the sender to try to unblind the real
+nodes and channels used. The sender can create onions with increased fees/cltv for the first
+blinded hop, starting with very low values. While the fee/cltv is below the real fee of the first
+hop, the sender will get an error from `N(0)`. Once the fee/cltv proposed actually satisfies the
+first hop's requirements, the error will come from another node `N(i)` inside the blinded path.
+
+The sender can then unblind channels one-by-one by discovering their real fees/cltv and matching
+those to existing channels in the graph.
+
+To mitigate this, when nodes along the blinded path are offered an invalid HTLC, they should:
+
+* return a dummy error encrypted with a throw-away key: the sender will receive an error she can't
+  decrypt and doesn't know which node generated it
+* hold the HTLC for a random amount of time before sending the error (otherwise the sender can
+  still use timing to guess what node errored out)
+
+Even with such mitigations the sender can discover the real fees/cltv of one of the blinded
+channels. To do so she uses the correct fees and cltv for all but one channel, and for that target
+channel she tries fees/cltv until the payment succeeds. Once the payment succeeds she knows the
+approximate fees/cltv of the target channel (but since the payment succeeded, she can't continue
+probing).
+
+Maybe nodes along the blinded path could use slightly different fees/cltv than what they publicly
+advertize? Or the recipient could add some fuzzing to it to blind them more?
+
+Are those mitigations enough? Or can a clever attacker still work around them?
```

Comment:
What prevent a gradually deanonymization attack ? Let's say you have blinded route:
           (E(0), fees(0), cltv(0), encrypted_blob(0))
           (B(1), fees(1), cltv(1), encrypted_blob(1))
           (B(2), fees(2), cltv(2), encrypted_blob(2))

Attacker can guess link E(0)-B(1) by fuzzing fees(0)/cltv(0) by iterating the right "key" fees/cltv based on the ones gossiped ? And use wacky values for fees(1)/cltv(1) to be sure to get an error at link, some kind of progressive fault-injection. Even if cltv_delta may be the same per-implementation, I fear than fees being based on node fee estimator, they are going to uniquely identify each link ?

Note, deanonymization wouldn't work if all final links are private, you don't have comparison point.

Reply:
I may be missing something, but I don't think this attack work. The attacker is only able to successfully fuzz one channel's fee/cltv, because once he guesses right the payment settles.

Reply:
Let's say beyond introduction point, you have 3 more hops, Bob, Caroll, and Dave.

On Bob-Caroll link you try with multiple public (fees,cltv) pair until you find one corresponding to an introduction point channel. But for Caroll-Dave link you send a 0-cltv_delta therefore getting a `incorrect_cltv_expiry`?

(maybe I should do ascii diagrams to describe attacks too ahah..)

Reply:
> you try with multiple public (fees,cltv) pair until you find one corresponding to an introduction point channel

Can you detail how the attacker knows he has found a valid channel?
The only way I currently see is that the payment goes through, so once the attacker has found one channel this way the payment will be fulfilled, so he doesn't get another shot at de-anonymizing more of the route.

I'll describe this further along with the uniform fee/cltv proposal, hopefully that will make it clearer.